runtime/libia2/threads.c

```diff
@@ -96,3 +94,100 @@ int __wrap_pthread_create(pthread_t *restrict thread,
   thread_thunk->data = data;
   return __real_pthread_create(thread, attr, ia2_thread_begin, thread_thunk);
 }
+
+// It's much simpler to only support a static number of created threads,
+// especially because we want to have very few dependencies.
+// If a program needs more threads, you can just increase this number.
+#define MAX_THREADS_SUPPORTED 4096
+
+struct ia2_all_threads_data {
+  pthread_mutex_t lock;
+  size_t num_threads;
+  pid_t tids[MAX_THREADS_SUPPORTED];
+  struct ia2_thread_data ia2_thread_data[MAX_THREADS_SUPPORTED];
+};
+
+#define array_len(a) (sizeof(a) / sizeof(*(a)))
+
+struct ia2_thread_data *ia2_all_threads_data_lookup(struct ia2_all_threads_data *const this, const pid_t tid) {
+  struct ia2_thread_data *data = NULL;
+  if (pthread_mutex_lock(&this->lock) == -1) {
```

Comment:
What is `-1` here and should we check for the other error codes?

Reply:
It just says it returns -1 on error like a lot of libc functions do.

Reply:
Fixed in ce350b114.  I misread the man page.  It says it returns 0 on success, so I'm just checking `!= 0` now.